### ISSUE TITLE
How do I debug issues related to auto-rotation?

### ISSUE THREAD

**Issue by wimo7083** [GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp#156]
### TL;DR

I have configured `secrets-store-csi-driver` to enable autorotation. When I update the `SecretProviderClass` to a new version of the GCP secret and apply the manifest I wait much longer than `--rotation-poll-interval=10s` and the `SecretProviderClassPodStatus` does not update.

**Expected behavior**

After some period the `SecretProviderClassPodStatus` updates and the secret (file mounted using volume in a pod) updates.

**Observed behavior**

`SecretProviderClassPodStatus` does not update and the secret (file mounted using volume in a pod) does not update.

### Reproduction

Note, custom-project is not the actual project name. If desired I can provide (through DM) access to a cluster in which this issue can be replicated.

1.  Following this example: https://github.com/boredabdel/gke-secret-manager/tree/main/hello-secret-csi-driver.

Generated these manifests
```bash
$ cat test.yaml
---
apiVersion: v1
kind: ServiceAccount
automountServiceAccountToken: true
metadata:
  annotations:
    iam.gke.io/gcp-service-account: testing-secrets@custom-project.iam.gserviceaccount.com
  name: testing-secrets
  namespace: default
---
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: app-secrets
  namespace: default
spec:
  provider: gcp
  parameters:
    secrets: |
      - resourceName: "projects/custom-project/secrets/test/versions/2"
        fileName: "good1.txt"
---
apiVersion: v1
kind: Pod
metadata:
  name: mypod
  namespace: default
spec:
  serviceAccountName: testing-secrets
  containers:
  - image: gcr.io/google.com/cloudsdktool/cloud-sdk:slim
    name: mypod
    resources:
      requests:
        cpu: 100m
    tty: true
    volumeMounts:
      - mountPath: "/var/secrets"
        name: mysecret
  volumes:
  - name: mysecret
    csi:
      driver: secrets-store.csi.k8s.io
      readOnly: true
      volumeAttributes:
        secretProviderClass: "app-secrets"
```

2.  Ran `kubectl apply -f test.yaml`
3.  After pod successful:
```bash
$ kubectl exec -it mypod -- cat /var/secrets/good1.txt
new-secret
```
4. update version to 3
```bash
$ sed -i 's|versions/2|versions/3|g' test.yaml
```
5 applied files `kubectl apply -f test.yaml`
6. Wait some period longer than 10s

```bash
$ k get SecretProviderClassPodStatus mypod-default-app-secrets -o json | jq '.status'
{
  "mounted": true,
  "objects": [
    {
      "id": "projects/custom-project/secrets/test/versions/2",
      "version": "projects/111111111111/secrets/test/versions/2"
    }
  ],
  "podName": "mypod",
  "secretProviderClassName": "app-secrets",
  "targetPath": "/var/lib/kubelet/pods/985a28cf-de16-4c33-a0e8-949e574eb99d/volumes/kubernetes.io~csi/mysecret/mount"
}
$ k get SecretProviderClass app-secrets -o json | jq '.spec.parameters'
{
  "secrets": "- resourceName: \"projects/custom-project/secrets/test/versions/3\"\n  fileName: \"good1.txt\"\n"
}
```

**Environment**

- OS: ubuntu bionic (local kubectl machine)
- Version: see below

```bash
$ gcloud container clusters list --project custom-project
NAME                    LOCATION     MASTER_VERSION    MASTER_IP      MACHINE_TYPE   NODE_VERSION      NUM_NODES  STATUS
shhhh us-central1  1.20.10-gke.1600  35.224.247.26  n1-standard-4  1.20.10-gke.1600  6          RUNNING
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.9", GitCommit:"9dd794e454ac32d97cde41ae10be801ae98f75df", GitTreeState:"clean", BuildDate:"2021-03-18T01:09:28Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20+", GitVersion:"v1.20.10-gke.1600", GitCommit:"ef8e9f64449d73f9824ff5838cea80e21ec6c127", GitTreeState:"clean", BuildDate:"2021-09-06T09:24:20Z", GoVersion:"go1.15.15b5", Compiler:"gc", Platform:"linux/amd64"}
```

Relevant container images:
```bash
us-docker.pkg.dev/secretmanager-csi/secrets-store-csi-driver-provider-gcp/plugin@sha256:cb491d4af4776ac352aac415676918fa7cd4ef1671e71c579175ef3e2820af09
k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.3.0
k8s.gcr.io/csi-secrets-store/driver:v1.0.0
k8s.gcr.io/sig-storage/livenessprobe:v2.4.0
```

**Additional information**

Honestly, just want to know how I should debug this and what I should be looking for. Thank you for your time and this nifty bit of software. I have also tried `version/latest`. The only way I've found to update is by deletion of the pod and recreation.

**Comment by wimo7083** [comment]
Thank you for the response. I did verify that I set the argument you suggested, and found a resolution.

Confirming that the flag is set:
```bash
$ k get ds -n kube-system csi-secrets-store             -o json  | jq '.spec.template.spec.containers[1] | .name, .args'
"secrets-store"
[
  "--endpoint=$(CSI_ENDPOINT)",
  "--nodeid=$(KUBE_NODE_NAME)",
  "--provider-volume=/etc/kubernetes/secrets-store-csi-providers",
  "--metrics-addr=:8095",
  "--enable-secret-rotation=true",
  "--rotation-poll-interval=10s",
  "--filtered-watch-secret=true",
  "--provider-health-check=false",
  "--provider-health-check-interval=2m"
]
```

Upon inspection of the logs I noticed that the permissions applied were lacking. 
```bash
E1208 22:25:55.596204       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.22.0/tools/cache/reflector.go:167: Failed to watch *v1.Secret: failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:kube-system:secrets-store-csi-driver" cannot list resource "secrets" in API group "" at the cluster scope
```

Giving the service account cluster-admin:
```bash
kubectl create clusterrolebinding tmp-cluster-admin --clusterrole=cluster-admin --serviceaccount=kube-system:secrets-store-csi-driver
```


BTW this is what permissions the service account had:

```bash
$ rbac-lookup | grep secrets-store-csi-driver
kube-system:secrets-store-csi-driver                    cluster-wide   ClusterRole/cluster-admin
kube-system:secrets-store-csi-driver                    cluster-wide   ClusterRole/secretproviderclasses-role
kube-system:secrets-store-csi-driver-provider-gcp       cluster-wide   ClusterRole/secrets-store-csi-driver-provider-gcp-role
```

specified roles:

```bash
$ k get clusterrole cluster-admin -o json | jq '.rules'
[
  {
    "apiGroups": [
      "*"
    ],
    "resources": [
      "*"
    ],
    "verbs": [
      "*"
    ]
  },
  {
    "nonResourceURLs": [
      "*"
    ],
    "verbs": [
      "*"
    ]
  }
]
$ k get clusterrole secretproviderclasses-role -o json | jq '.rules'
[
  {
    "apiGroups": [
      ""
    ],
    "resources": [
      "events"
    ],
    "verbs": [
      "create",
      "patch"
    ]
  },
  {
    "apiGroups": [
      ""
    ],
    "resources": [
      "pods"
    ],
    "verbs": [
      "get",
      "list",
      "watch"
    ]
  },
  {
    "apiGroups": [
      "secrets-store.csi.x-k8s.io"
    ],
    "resources": [
      "secretproviderclasses"
    ],
    "verbs": [
      "get",
      "list",
      "watch"
    ]
  },
  {
    "apiGroups": [
      "secrets-store.csi.x-k8s.io"
    ],
    "resources": [
      "secretproviderclasspodstatuses"
    ],
    "verbs": [
      "create",
      "delete",
      "get",
      "list",
      "patch",
      "update",
      "watch"
    ]
  },
  {
    "apiGroups": [
      "secrets-store.csi.x-k8s.io"
    ],
    "resources": [
      "secretproviderclasspodstatuses/status"
    ],
    "verbs": [
      "get",
      "patch",
      "update"
    ]
  }
]
```

**Comment by tam7t** [comment]
Ahhhh, at some point the rotation feature began needing secret permissions:

https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/main/deploy/rbac-secretproviderrotation.yaml

https://secrets-store-csi-driver.sigs.k8s.io/getting-started/installation.html#alternatively-deployment-using-yamls

I think that:

```
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/secrets-store-csi-driver/main/deploy/rbac-secretprovidersyncing.yaml
```

should install the required permissions

**Comment by tam7t** [comment]
I need to come back to this tomorrow to look into why our e2e tests pass: https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp/blob/5803f528ee9772180fb26103e3a505882563e720/test/e2e/mount_test.go#L364

when I dont believe that we ever apply `rbac-secretproviderrotation.yaml` permissions...

**Comment by tam7t** [comment]
Tested manually on a 1.21 cluster, installed v1.0.0 of driver and provider:

```
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/secrets-store-csi-driver/v1.0.0/deploy/rbac-secretproviderclass.yaml
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/secrets-store-csi-driver/v1.0.0/deploy/rbac-secretprovidersyncing.yaml
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/secrets-store-csi-driver/v1.0.0/deploy/csidriver.yaml
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/secrets-store-csi-driver/v1.0.0/deploy/secrets-store.csi.x-k8s.io_secretproviderclasses.yaml
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/secrets-store-csi-driver/v1.0.0/deploy/secrets-store.csi.x-k8s.io_secretproviderclasspodstatuses.yaml
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/secrets-store-csi-driver/v1.0.0/deploy/secrets-store-csi-driver.yaml
kubectl apply -f https://raw.githubusercontent.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp/v1.0.0/deploy/provider-gcp-plugin.yaml
```

enabled rotation with:

```
KUBE_EDITOR="gsed -i 's/enable-secret-rotation=false/enable-secret-rotation=true/g;s/rotation-poll-interval=2m/rotation-poll-interval=10s/g'" kubectl edit daemonset csi-secrets-store -n kube-system
```

I believe the integration tests pass without `rbac-secretproviderrotation.yaml ` applied because `secretprovidersyncing-rolebinding` clusterRoleBinding contains a _superset_  of the `secretproviderrotation-rolebinding` permissions

**Comment by wimo7083** [comment]
Thank you for the update. Since this is an issue with the upstream provider's documentation is there anything you would like to see other than filing an issue with them?
